src/raft.rs

```diff
@@ -484,19 +484,56 @@ impl<T: Storage> Raft<T> {
         m.set_entries(RepeatedField::from_vec(ents));
         m.set_commit(self.raft_log.committed);
         if !m.get_entries().is_empty() {
-            match pr.state {
-                ProgressState::Replicate => {
-                    let last = m.get_entries().last().unwrap().get_index();
-                    pr.optimistic_update(last);
-                    pr.ins.add(last);
+            let last = m.get_entries().last().unwrap().get_index();
+            self.update_progress_state(last, pr);
+        }
+    }
+
+    fn update_progress_state(&mut self, last: u64, pr: &mut Progress) {
+        match pr.state {
+            ProgressState::Replicate => {
+                pr.optimistic_update(last);
+                pr.ins.add(last);
+            }
+            ProgressState::Probe => pr.pause(),
+            _ => panic!(
+                "{} is sending append in unhandled state {:?}",
+                self.tag, pr.state
+            ),
+        }
+    }
+
+    fn try_batching(
+        &mut self,
+        to: u64,
+        pr: &mut Progress,
+        ents: &mut Vec<Entry>,
+        term: u64,
+    ) -> (bool, bool, u64) {
```

Comment:
can you explain the result tuple here?

Reply:
First one says whether it is batched or not. (used to create new msg if not matched)
The second one is to say whether the entries in the rpc are empty or not (used whether to update the progress state or not)
the third one is the last index of the entry(index to be updated).

I think I can move the last two step inside this function itself. your views?

Reply:
Hi @sch00lb0y 

Sorry for the late reply, I prefer keeping the return values simply. If you can move them, it is better. If not, you can add comments for the return values.

Reply:
@siddontang I've moved `update_progress_state` from `send_append` to `try_batching`